### PR TITLE
x86: grow rootfs_data partition before format

### PR DESCRIPTION
**Pull request by luizluca**
The non-default f2fs-tools can expand a f2fs only when not in use.
/overlay is only unmounted while in failsafe, which also prevents the
use of any installed packages. Today, the easier way is to resize the
partition after installation is resizing the partition and use failsafe
to reformat the /overlay, requiring physical device access. And, after
an upgrade, the partition is reverted.

This patch adds a grow_rootfspart shellscript partitioner that expands
the seconds MBR or GPT partition until the end of the device. It can
run directly to an img file or from the installed device. It only requires
standard Linux commands already available on default x86 images. After
the partition resizes, it will refresh partition information in kernel.

For GPT, it will always touch the disk at least once as the current
generated images do not use all the available space and it also misses
the GPT backup header at the end of the disk.

preinit/43_grow_rootfspart will run it when it does not detect an f2fs/ext4
filesystem at the overlay position and only when not in failsafe. That
condition is also met during the first boot after an upgrade. mount_root
will format rootfs_data alter that.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

### CODE REVIEW

**Comment by luizluca** [comment]
Today it is a standalone script but many functions used might have synergy with those in /lib/upgrade/

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	rootfs_data_magic=$(hexdump -s $((squashfs_size+0x400)) -n 4 -e '1/4 "%08x" "\n"' "${diskdev}2")
+	[ "$rootfs_data_magic" != "f2f52010" ] || return # F2FS
+	rootfs_data_magic=$(hexdump -s $((squashfs_size+0x438)) -n 2 -e '1/2 "%04x" "\n"' "${diskdev}2")
+	[ "$rootfs_data_magic" != "ef53" ] || return # ext3
+	/usr/sbin/grow_rootfspart "${diskdev}"
+    fi 
+}
+
+boot_hook_add preinit_main run_grow_root_part
+
```
Comment (adschm) [inline target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:34]: empty line at EOF can be dropped.

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#!/bin/sh
```
Comment (adschm) [inline target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:1]: Preinit is sourced (/lib !), please remove shebang and execute bit on file

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+get_squashfs_size() {
+	local low high
+	low=$(hexdump -s 40 -n 4 -e '1/4 "%08x" "\n"' "$1")
+	high=$(hexdump -s 44 -n 4 -e '1/4 "%08x" "\n"' "$1")
+	echo $((0x$high$low))
+}
+
+run_grow_root_part() {
+    [ "$FAILSAFE" != "true" ] || return
```
Comment (adschm) [inline target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:11]: [ "$FAILSAFE" = "true" ] && return
?

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    local diskdev squashfs_magic
+    local SQUASHFS_ALIGN=$((0x10000))
+
+    . /lib/upgrade/common.sh
+
+    if export_bootdevice && export_partdevice diskdev 0; then
+	diskdev=/dev/$diskdev
```
Comment (adschm) [inline target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:19]: Your indent is somewhat arbitrary. Please check with some similar files and try to use it consistent to those (as far as possible). I'd just use tabs here ...

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        squashfs_magic=$(dd bs=4 count=1 if="${diskdev}2" 2>/dev/null)
+	[ "$squashfs_magic" = "hsqs" ] || return
+	squashfs_size=$(get_squashfs_size "${diskdev}2")
+	squashfs_size=$((squashfs_size + (SQUASHFS_ALIGN-1) & (~(SQUASHFS_ALIGN-1))))
+	[ "$squashfs_size" -gt 0 ] || return
+	rootfs_data_magic=$(hexdump -s $((squashfs_size+0x400)) -n 4 -e '1/4 "%08x" "\n"' "${diskdev}2")
+	[ "$rootfs_data_magic" != "f2f52010" ] || return # F2FS
```
Comment (adschm) [inline target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:26]: Again, [ = ] && return ?
Same two lines below.

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+crc32_hex(){
+	gzip -c | tail -c8 | hexdump -n4 -e '"%.8x" "\n"'
+}
+
+
+msg(){
+        printf "grow_rootfspart: %s\n" "$*" >&2
+
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:42]: Indent not matching the rest, unneeded empty line.

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+}
+die() {
+	local err=$1; shift
+	msg "$@"
+	exit "$err"
+}
+
+# swap endianness
+bswap(){
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:51]: I'd prefer to have a space between () and {
This is currently mixed in the file ...

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+	b3=$(dec2hex "$((c & 0xff))" 1)
+	printf "%s\n" "$b1$b2$b3"
+}
+
+mbr_parse_part() {
+	local partentry_hex=$1
+	#mbr_status=${partentry_hex:MBR_PARTSTATUS*2:MBR_PARTSTATUS_SIZE*2}
+	#mbr_first_chs=${partentry_hex:MBR_PARTFIRST_CHS*2:MBR_PARTCHS_SIZE*2}
+	#mbr_type=${partentry_hex:MBR_PARTTYPE*2:MBR_PARTTYPE_SIZE*2}
+	#mbr_last_chs=${partentry_hex:MBR_PARTLAST_CHS*2:MBR_PARTCHS_SIZE*2}
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:83]: Do we need these comments?

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+
+        gpt_partition_type=${partentry_hex:GPT_PARTITION_TYPE*2:GPT_PARTITION_TYPE_SIZE*2}
+        [ "$gpt_partition_type" != "00000000000000000000000000000000" ] || return 1
+
+        gpt_partition_first_lba_hex_le=${partentry_hex:GPT_PARTITION_FIRST_LBA*2:GPT_LBA_SIZE*2}
+        gpt_partition_first_lba_hex=$(bswap "$gpt_partition_first_lba_hex_le")
+        gpt_partition_last_lba_hex_le=${partentry_hex:GPT_PARTITION_LAST_LBA*2:GPT_LBA_SIZE*2}
+        gpt_partition_last_lba_hex=$(bswap "$gpt_partition_last_lba_hex_le")
+        #gpt_partition_attribute_hex_le=${partentry_hex:GPT_PARTITION_ATTRIBUTE*2:GPT_PARTITION_ATTRIBUTE_SIZE*2}
+        #gpt_partition_label_utf16le=${partentry_hex:GPT_PARTITION_LABEL*2:GPT_PARTITION_LABEL_SIZE*2}
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:101]: Do we need these comments?

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+MBR_PART1_ENTRY=$((0x01BE))
+MBR_PART2_ENTRY=$((0x01CE))
+MBR_PART3_ENTRY=$((0x01DE))
+MBR_PART4_ENTRY=$((0x01EE))
+MBR_PARTENTRY_SIZE=16
+
+#MBR_PARTSTATUS=$((0x0)) MBR_PARTSTATUS_SIZE=1
+MBR_PARTFIRST_CHS=$((0x01)) MBR_PARTLAST_CHS=$((0x05)) MBR_PARTCHS_SIZE=3
+#MBR_PARTTYPE=$((0x04)) MBR_PARTTYPE_SIZE=1
+MBR_PARTFIRST_SECTOR=$((0x08)) MBR_PARTNUMBER_OF_SECTOR=$((0x0C)) MBR_PARTSECTOR_SIZE=4
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:121]: Do we need these comments?

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+GPT_CURRENT_LBA=$((0x18)) GPT_BACKUP_LBA=$((0x20)) GPT_LBA_SIZE=8
+GPT_FIRST_LBA=$((0x28)) GPT_LAST_LBA=$((0x30))
+GPT_PARTITIONS_LBA=$((0x48))
+GPT_NUMBER_OF_PARTITIONS=$((0x50)) GPT_NUMBER_OF_PARTITIONS_SIZE=4
+GPT_PARTITION_ENTRY_SIZE=$((0x54)) GPT_PARTITION_ENTRY_SIZE_SIZE=4
+
+GPT_PARTITION_TYPE=$((0x0)) GPT_PARTITION_TYPE_SIZE=16
+GPT_PARTITION_FIRST_LBA=$((0x20)) GPT_PARTITION_LAST_LBA=$((0x28))
+#GPT_PARTITION_ATTRIBUTE=$((0x38)) GPT_PARTITION_ATTRIBUTE_SIZE=8
+#GPT_PARTITION_LABEL=$((0x38)) GPT_PARTITION_LABEL_SIZE=72
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:141]: Do we need these comments?

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+
+disk=$1
+disk=$(readlink -f "$1")
+
+[ -r "$1" ] ||
+	die 1 "Failed to read '$1'"
+
+if [ -b "$disk" ]; then
+        disk_major_minor=$(ls -l "$disk" | awk '{ print substr($5,1,length($5)-1) ":" $6 }')
+        disk_size_sectors=$(($(cat "/sys/dev/block/$disk_major_minor/size")))
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:154]: From this point, indent start to be broken regularly again.

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    [ "$FAILSAFE" != "true" ] || return
+
+    local diskdev squashfs_magic
+    local SQUASHFS_ALIGN=$((0x10000))
+
+    . /lib/upgrade/common.sh
+
+    if export_bootdevice && export_partdevice diskdev 0; then
+	diskdev=/dev/$diskdev
+        squashfs_magic=$(dd bs=4 count=1 if="${diskdev}2" 2>/dev/null)
```
Comment (wulfy23) [inline target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:20]: logic wont handle mmc devs ( = /dev/mmcblk02 != /dev/mmcblk0p2 )

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+	local buf=$1
+	local seek=$2
+	local modified=$3
+	printf "%s" "${buf:0:seek*2}"
+	printf "%s" "$modified"
+	printf "%s\n" "${buf:seek*2+${#modified}}"
+}
+
+crc32_hex(){
+	gzip -c | tail -c8 | hexdump -n4 -e '"%.8x" "\n"'
```
Comment (wulfy23) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:36]: consider making this '/bin/busybox gzip ...' for robustness

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 2 above)
Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:1]: It was copied from another preinit file:

```
root@(none):/lib/preinit# head -1 * | grep '#!/bin/sh' | wc -l
13
```

Maybe they all should be removed.

Done

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 3 above)
Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:11]: I'm skipping the growfs if failsafe was trigged. I copied from 40_run_failsafe_hook.
Should I check for /tmp/.failsafe instead?

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 4 above)
Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:19]: I replaced all to tabs. It got messes while editing both with vim-full and busybox vim.

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 5 above)
Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:26]: I normally program in a way that every command will not return false if it is not an error. That allows me to use "set -e" or 'trap "xxx" ERR'. But, if it is not clear, I can invert the logic to == && or simply use a verbose "if then". Which one would be better?

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 6 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:42]: Indent looks good now.

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 7 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:51]: spaces added

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 8 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:83]: No. Most of them where used only during the debug phase. I can remove all commented code. I kept them intentionally if someone might be interested on copying them to a library (like /lib/upgrade/common.sh).

Should I remove them?

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 9 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:101]: Same as before, no.

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 11 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:141]: Same as before, no

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 12 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:154]: It is fixed now.

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 13 above)
Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:20]: I didn't think about mmc devices for x86. We also have nvm (nvm0n0p2). Fixed

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 14 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:36]: Both, busybox or standalone, will work as expected. If I add a busybox, it might fail when there is no busybox installed.
This script can be executed from the host modifying the img file, where busybox might not be available. I would keep it as it is.

Comment (wulfy23) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:36]: yeah... should be ok @ preinit... there is a bug at the moment @ stage2 + gzip full breaking zcat ( for devices using img.gz sysupgrade... agree that it likely wont effect this... ( it's really a stage2 issue / gzip full package issue ) but felt it would be good to mention it.

**Comment by luizluca** [comment]
I'll squash those commits after reviewed

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 2 above)
Comment (adschm) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart]: I did that a while ago but probably did not hit them all ...
https://github.com/openwrt/openwrt/commit/48c1fdd046f409fd5e28932f1d53343fa214fbc4

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 3 above)
Comment (adschm) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart]: I just wonder why this is "not equal"->OR where "equal"->AND looks simpler to me. Just about the logic inversion ...

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 8 above)
Comment (adschm) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart]: The script does what it does. I don't see much benefit by leftover debug code, even more since it's not commented.

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 3 above)
Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart]: I'll change it if it looks better.

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 8 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart]: I removed all of them

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 5 above)
Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart]: Changed all [!=] || to [=] &&

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 3 above)
Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart]: Changed all [!=] || to [=] &&

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 9 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart]: Commented code is gone

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 14 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:36]: Thanks for the mention. I guess sysupgrade is a much more complex env as it creates a ramfs from a running system.
If you have multiple implementations for a program, it might get the wrong file.

In preinit, even if gzip is builtin, it should not be a problem as both works as expected.

**Comment by luizluca** [comment]
I guess we have resolved all conversations.

@adschm @wulfy23 , it is ok to squash the commits?

If someone wants to test this, it is quite easy:

```
$ gunzip -kf bin/targets/x86/64/openwrt-x86-64-generic-squashfs-combined.img.gz ; truncate --size=4TiB bin/targets/x86/64/openwrt-x86-64-generic-squashfs-combined.img ; ./scripts/qemustart x86 64 -usb -m 1024 -enable-kvm --network --rootfs bin/targets/x86/64/openwrt-x86-64-generic-squashfs-combined.img
# gunzip -kf bin/targets/x86/64/openwrt-x86-64-generic-squashfs-combined-efi.img.gz ; truncate --size=4TiB bin/targets/x86/64/openwrt-x86-64-generic-squashfs-combined-efi.img ; ./scripts/qemustart x86 64 -usb -m 1024 -enable-kvm --network --rootfs bin/targets/x86/64/openwrt-x86-64-generic-squashfs-combined-efi.img
```

truncate can create large sparse disks. I tested up to my FS limit (16 TiB).
You can also compare with what parted does:

```
parted -s bin/targets/x86/64/openwrt-x86-64-generic-squashfs-combined.img resize 2 100%
# this might give some warns about missing backup GPT header
parted -s bin/targets/x86/64/openwrt-x86-64-generic-squashfs-combined-efi.img resize 2 100%
```

and the result from grow_rootfspart:

```
target/linux/x86/base-files/usr/sbin/grow_rootfspart bin/targets/x86/64/openwrt-x86-64-generic-squashfs-combined.img
target/linux/x86/base-files/usr/sbin/grow_rootfspart bin/targets/x86/64/openwrt-x86-64-generic-squashfs-combined-efi.img
```

It should match bit-to-bit.

/usr/sbin/grow_rootfspart can also be executed manually from failsafe with equivalent results (bit-to-bit).

This PR will turn CONFIG_TARGET_ROOTFS_PARTSIZE ("Root filesystem partition size") value mostly irrelevant.

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+	if export_bootdevice && export_partdevice diskdev 0; then
+		diskdev=/dev/$diskdev
+		case "$diskdev" in
+			*[0-9]) part2dev=${diskdev}p2;;
+			*)      part2dev=${diskdev}2;;
+		esac
+		[ -e "$part2dev" ] || return
+		squashfs_magic=$(dd bs=4 count=1 if="$part2dev" 2>/dev/null)
+		[ "$squashfs_magic" = "hsqs" ] || return
```
Comment (wulfy23) [inline target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:24]: this fails on ext4 implementations... suggest a subroutine/function once you ( and/or TO + CASE ) identify the FSTYPE, cascaded logix seems problematic here...

Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:24]: This is intentional not executed for ext4 rootfs (although it will work as expected for squashfs+ext4). Growing part2 is only required at preinit if it will be used by an overlay filesystem that will be formatted a little bit after. I just know the squashfs case. Would it be anything else similar to squashfs+overlay case for x86 targets that I should consider?

ext4 should be expandable, even while in use. It should only require resize2fs installed. There is no advantage of doing it during pre-init as ext4 is already mounted.

The script grow_rootfspart will gladly expand part2 if execute manually, even refreshing partitions without a reboot:

root@OpenWrt:/# /usr/sbin/grow_rootfspart /dev/vda
grow_rootfspart: Growing /dev/vda partition 2 from 212992 to 433152 sectors
grow_rootfspart: Saving /dev/vda new MBR partition table...

I could save the user from installing a partitioner. However, the ext4 shipped with a ext4 flavor does not seem to like being expanded. It might be caused by a fine-tune optimization trying to reduce the ext4 disk usage:

```
root@OpenWrt:/# opkg install resize2fs
root@OpenWrt:/# resize2fs /dev/vda2 
resize2fs 1.45.6 (20-Mar-2020)
Filesystem at /dev/vda2 is mounted on /; on-line resizing required
old_desc_blocks = 1, new_desc_blocks = 1
[   60.123874] EXT4-fs (vda2): resizing filesystem from 32768 to 54144 blocks
[   60.126318] EXT4-fs warning (device vda2): reserve_backup_gdb:1035: reserved block 8 not at offset 7
[   60.128802] EXT4-fs warning (device vda2): ext4_resize_fs:2122: error (-22) occurred during file system resize
[   60.130514] EXT4-fs (vda2): resized filesystem to 32768
Performing an on-line resize of /dev/vda2 to 54144 (4k) blocks.
[   60.133087] EXT4-fs warning (device vda2): reserve_backup_gdb:1035: reserved block 8 not at offset 7
resize2fs: Invalid argument While trying to add group #1
```

As long as this persists, it is a dead end. If it could be expanded, I would simply run the `grow_rootfspart /dev/xxx2` and `resize2fs /dev/xxx2` after the system was up.

Comment (wulfy23) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:24]: ok... so in it's current form... you expect it to fail/end silently on ext4 implementations... ( the mentioned code is not silent produces errors )...

in this case... maybe clarify this in the pr description and/or consider some way to toggle this functionality, either as a package selection / Makefile@squasfs/f2fs/pre-init option etc.

Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:24]: > ok... so in it's current form... you expect it to fail/end silently on ext4 implementations...
Yes

> in this case... maybe clarify this in the pr description and/or consider some way to toggle this functionality, either as a package selection / Makefile@squasfs/f2fs or pre-init option etc.

This feature is specific for x86 targets where storage normally is not a big issue. I could convert it to a package but it would be mostly useless for other targets. No other target appears to be using GPT (even less likely MBR) for booting.The preinit part is intentionally innocuous for ext4 rootfs but grow_rootfspart could still be usable.

It would probably use a image config value (dependent on x86) instead of a standalone package. I just don't know if it is worth it adding a new menu entry and extra image building logic. Anyway, I can do it without a problem. However, before that, I just need to know if there is interest to merge this PR.

**Comment by luizluca** [comment]
@adschm @wulfy23 , is there interest in merging this PR?

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 15 above)
Comment (wulfy23) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:24]: most sdcard images are also capable / compatible with this codeset ( rpi, bpi, r2s etc. )

another possibility is a "sysupgrade -E" ( extend ) style option that would be silently ignored on typical non-extendable/flash devices...

it's part of a larger push on such devices ( multiboot pr, gpt, efi etc. even dual-fw to some extent ) to extensify disk operations and carries with it complexities which overlap with typical logic... and bring compound implications when it comes to image consistency and predictability across the project... ( dual-fw logic @ advanced reboot app is virtually hardcoded and needs to be library-fied )

i'm unable to speak for merging... but I would like to see the majority of the options and functions discussed condensed and selectable within the build system... or via clear user hooks and parameters...

adrians upgrade json parameters are a nice clear window into one place some of these capabilities might be 'negotitated'/hooked into.... ( note: on a currently running device, there is no way to query "am I factory-squashfs" or "am I factory-ext4" the asu software for example just selects the first image returned by the server... you have an (on-device) fix for this... so it should be in common.sh or common-fulldisk.sh etc. etc. and/or be available via os-release or json call device partitions etc. etc. if it's in os-release/system from factory... then you can do away with alot of validation / determinism )

in short, would I like to see it merged? > yes... with user selectability and better library-fication...

not consolidating alot of this stuff now impedes extensification and leads to drama down the road...

my 2c

**Comment by adschm** [comment]
@luizluca Please squash your changes into one (or a few) proper commits.

**Comment by luizluca** [comment]
> @luizluca Please squash your changes into one (or a few) proper commits.

Squashed.

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart [hunk 16]
```diff
@@ -0,0 +1,36 @@
+get_squashfs_size() {
```
Comment (adschm) [inline target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:1]: Please add SPDX license identifiers on all new files.

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 17]
```diff
@@ -0,0 +1,320 @@
+#!/bin/sh
+# shellcheck disable=SC2039
+#
+# Grow partition 2 (rootfs_data) until the end of the disk
+#
+#
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:6]: Extra comment lines may be dropped.

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+	printf "%s\n" "$1" | sed -r 's/(.{2})/\\\\x\1 /g' | xargs printf '%b' 
+}
+
+writehex2bin() {
+	local output=$1
+	local seek=$2
+	local hexbuf=$3
+	local length=$((${#hexbuf}/2))
+	hex2bin "$hexbuf" | dd bs=1 seek="$seek" count="$length" of="$output" conv=notrunc 2>/dev/null
+}
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:25]: missing empty line.

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+	printf "%s" "${buf:0:seek*2}"
+	printf "%s" "$modified"
+	printf "%s\n" "${buf:seek*2+${#modified}}"
+}
+
+crc32_hex() {
+	gzip -c | tail -c8 | hexdump -n4 -e '"%.8x" "\n"'
+}
+
+
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:39]: unneeded empty line.

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart [hunk 20]
```diff
@@ -0,0 +1,320 @@
...
+msg() {
+	printf "grow_rootfspart: %s\n" "$*" >&2
+
+}
```
Comment (adschm) [inline target/linux/x86/base-files/usr/sbin/grow_rootfspart:43]: missing empty line.

File: target/linux/x86/base-files/lib/preinit/43_grow_rootfspart (hunk 16 above)
Comment (luizluca) [reply target/linux/x86/base-files/lib/preinit/43_grow_rootfspart:3]: done

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 18 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:27]: done

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 20 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:45]: done

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 19 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart]: done

**Comment by luizluca** [comment]
All done and rebased on master

File: target/linux/x86/base-files/usr/sbin/grow_rootfspart (hunk 17 above)
Comment (luizluca) [reply target/linux/x86/base-files/usr/sbin/grow_rootfspart:8]: done

**Comment by adschm** [comment]
A Tested-by always helps. But note that I'm not dealing with x86, so you have to wait for somebody else nevertheless.

**Comment by luizluca** [comment]
To keep this PR alive, I just rebased and removed some extra spaces before \n. It still works as expected.

**Comment by aparcar** [comment]
@dangowrt didn't you develop a package for that?

**Comment by dangowrt** [comment]
I've developed `autopart` [1] which does something slightly different (and imho much more reasonable): It assigns all the remaining space on the block device which was used to boot from to an LVM2 physical volume.

`rootfs_data` is meant for configuration and to install additional software packages -- it is very hard to imaging how this kind of use could occupy more than 1GB, leave alone TB...

Growing `rootfs_data` beyond even the size of the RAM might seem compelling, but actually doesn't make much sense:
 - `rootfs_data` is wiped and re-created every time you do sysuprade
    - configuration has to survive in a tarball which is **created in tmpfs**
    - (in case of x86) then moved to `/boot`
 - the most space hungry application (container images) doesn't actually work on top of an overlayfs
 - apart from being technically challenging it would also be silly to copy and re-write user data (think: samba share content, torrent downloads, ...) every time `sysupgrade` is used

So while it is of course easily possible to auto-expand rootfs_data to the whole disk, that actually creates more problems than it solves.

Edit: editing the partition table using on-board busybox tools is cool, of course ;)

[1]: https://github.com/openwrt/packages/blob/master/utils/uvol/files/autopart.defaults
